### ISSUE TITLE
Performance Optimizations for IndexHierarchy set operations

### ISSUE THREAD

**Issue by chaburkland** [static-frame/static-frame#625]
**Example**

```python
import typing as tp

import numpy as np
import static_frame as sf

def index_hierarchy_intersection(
    lhs: sf.IndexHierarchy, rhs: sf.IndexHierarchy
) -> sf.IndexHierarchy:

    # ~2%
    remapped_indexers_lhs: tp.List[np.ndarray] = []
    remapped_indexers_rhs: tp.List[np.ndarray] = []
    num_unique_elements_per_depth: tp.List[int] = []

    for idx_lhs, idx_rhs, indexer_lhs, indexer_rhs in zip(
        lhs._indices, rhs._indices, lhs._indexers, rhs._indexers
    ):
        union = idx_lhs.union(idx_rhs)
        num_unique_elements_per_depth.append(len(union))

        remapped_a = idx_lhs._index_iloc_map(union)
        remapped_b = idx_rhs._index_iloc_map(union)

        remapped_indexers_lhs.append(remapped_a[indexer_lhs])
        remapped_indexers_rhs.append(remapped_b[indexer_rhs])

    (
        bit_offset_encoders,
        encoding_can_overflow,
    ) = sf.core.loc_map.HierarchicalLocMap.build_offsets_and_overflow(
        num_unique_elements_per_depth=num_unique_elements_per_depth
    )

    encoding_dtype = object if encoding_can_overflow else np.uint64

    # ~4%
    def build_hashes(arrs: tp.List[np.ndarray]) -> np.ndarray:
        arr = np.array(arrs, dtype=encoding_dtype).T
        arr <<= bit_offset_encoders
        return np.bitwise_or.reduce(arr, axis=1)

    lhs_hashes = build_hashes(remapped_indexers_lhs)
    rhs_hashes = build_hashes(remapped_indexers_rhs)

    # ~53%
    mask = np.in1d(lhs_hashes, rhs_hashes)

    # ~2%
    blocks = lhs._blocks._extract(mask)

    # ~39%
    return sf.IndexHierarchy._from_type_blocks(
        blocks,
        name=lhs.name,
        own_blocks=True,  # We are not slicing, so this necessarily will copy. I think? @flexatone
        index_constructors=lhs._index_constructors,
    )


def index_hierarchy_union(
    lhs: sf.IndexHierarchy, rhs: sf.IndexHierarchy
) -> sf.IndexHierarchy:
    # ~0%
    remapped_indexers_lhs: tp.List[np.ndarray] = []
    remapped_indexers_rhs: tp.List[np.ndarray] = []
    num_unique_elements_per_depth: tp.List[int] = []

    for idx_lhs, idx_rhs, indexer_lhs, indexer_rhs in zip(
        lhs._indices, rhs._indices, lhs._indexers, rhs._indexers
    ):
        union = idx_lhs.union(idx_rhs)
        num_unique_elements_per_depth.append(len(union))

        remapped_a = idx_lhs._index_iloc_map(union)
        remapped_b = idx_rhs._index_iloc_map(union)

        remapped_indexers_lhs.append(remapped_a[indexer_lhs])
        remapped_indexers_rhs.append(remapped_b[indexer_rhs])

    (
        bit_offset_encoders,
        encoding_can_overflow,
    ) = sf.core.loc_map.HierarchicalLocMap.build_offsets_and_overflow(
        num_unique_elements_per_depth=num_unique_elements_per_depth
    )

    encoding_dtype = object if encoding_can_overflow else np.uint64

    # ~3%
    def build_hashes(arrs: tp.List[np.ndarray]) -> np.ndarray:
        arr = np.array(arrs, dtype=encoding_dtype).T
        arr <<= bit_offset_encoders
        return np.bitwise_or.reduce(arr, axis=1)

    lhs_hashes = build_hashes(remapped_indexers_lhs)
    rhs_hashes = build_hashes(remapped_indexers_rhs)

    # ~46%
    unique_to_b = np.in1d(rhs_hashes, lhs_hashes, invert=True)

    # ~2%
    lhs_blocks = lhs._blocks
    rhs_blocks = rhs._blocks._extract(unique_to_b)
    blocks = sf.TypeBlocks.from_blocks(
        sf.TypeBlocks.vstack_blocks_to_blocks((lhs_blocks, rhs_blocks))
    )

    # ~47%
    return sf.IndexHierarchy._from_type_blocks(
        blocks,
        name=lhs.name,
        own_blocks=True,  # We are not slicing, so this necessarily will copy. I think? @flexatone
        index_constructors=lhs._index_constructors,
    )

```

**Platform**
```python
<Series: platform>
<Index>
platform           Linux-5.4.0-126-generic-x86_64-with-glibc2.29
sys.version        3.8.12 (default, Aug 26 2022, 13:23:08) [GCC 9.4.0]
static-frame       0.9.14
numpy              1.19.2
pandas             0.25.3
xlsxwriter         1.1.5
openpyxl           3.0.9
xarray             <ModuleNotFoundError>
tables             3.6.1
pyarrow            0.17.0
msgpack            (1, 0, 0)
msgpack_numpy      None
<<U13>             <object>

```

**Comment by chaburkland** [comment]
@flexatone I have observed very significant performance improvements using these optimizations. Is it possible to integrate these somehow in `_index_many_to_one`? From what I can tell, it's essentially circumventing the set-like methods on the index objects themselves, and opting for operating on the underlying data directly.

This is proving to be very slow when working with `IndexHierarchies`, and my discovery of the alternative approach I posted above, ideally would lead to methods likes `sf.Frame.from_concat` becoming much faster.
